client/src/components/EmojiPicker.js

```diff
@@ -0,0 +1,169 @@
+'use strict';
+
+import _ from 'lodash';
+import React from 'react';
+import ReactEmoji from "react-emoji";
+import { emojiData } from "../utils/emojilist";
+import emojisAnnotations from 'emoji-annotation-to-unicode';
+import 'styles/EmojiPicker.scss';
+
+const modulo = (op1, op2) => (((op1 % op2) + op2) % op2);
+const supportedEmojis = Object.values(emojisAnnotations).filter((e) => e !== '1f642');
+const emojiList = _.pickBy(emojiData, (e) => supportedEmojis.indexOf(e.unicode) > -1);
+
+const filterEmojis = function(emojiList, word, amount) {
+  return Object.values(emojiList)
+    .filter((e) => e.shortname.indexOf(word) > -1 || e.aliases.indexOf(word) > -1 || word === '')
+    .slice(0, amount);
+};
+
+const emojify = (emojis, highlightedIndex, onMouseEnter, onClick) => {
+  const emojiOpts = { emojiType: 'emojione' };
+  return emojis
+    .map(e => ReactEmoji.emojify(e.shortname, emojiOpts))
+    .map((e, index) => <li key={index}
+                           onMouseEnter={ e => { e.preventDefault(); onMouseEnter(index); }}
+                           onClick={ e => { e.preventDefault(); onClick(index); }}
+                           className={index === highlightedIndex ? 'selected' : ''}>{e}</li>);
+};
+
+const EmojiList = ({
+    emojis,
+    highlightedIndex,
+    onMouseEnter,
+    onClick
+}) => <ul> { emojify(emojis, highlightedIndex, onMouseEnter, onClick) } </ul>
+
+class EmojiPicker extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            emojis: [],
+            highlightedIndex: 0,
+            keysDown: {}
+        }
+    }
+
+    selectEmoji(index) {
+        const emojiShortName = this.state.emojis[index] ? this.state.emojis[index].shortname : this.props.filterText;
+        this.props.onSelectEmoji(emojiShortName);
+    }
+
+    onKeyUp(event) {
+        if (event.which === 16) {
+            // Shift
+            const keysDown = { 'shift' : false };
+            this.setState({ keysDown: keysDown });
+        }
+    }
+    onKeyDown(event) {
+        if (event.which === 8) {
+            // Backspace
+            this.selectEmoji(-1);
+            if(this.state.highlightedIndex !== 0) {
+                event.preventDefault();
+                this.setState({ highlightedIndex: 0});
+            }
+        } else if (event.which === 16) {
+            // Shift
+            const keysDown = { 'shift' : true };
+            this.setState({ keysDown: keysDown });
+        } else if(event.which === 9) {
+            // Tab
+            event.preventDefault();
+            if (this.state.keysDown['shift']) {
```

Comment:
Instead of tracking the shift key press in the state, you can directly get it from the `event` passed to the onKeyDown. This should simplify the key check code a little bit (including the lines above).
```javascript
if (event.shiftKey) {
...
```